java/flight/flight-sql/src/test/java/org/apache/arrow/flight/sql/example/FlightSqlStatelessExample.java

```diff
@@ -0,0 +1,246 @@
+                (DoPutPreparedStatementResultPOJO) ois.readObject();
+        final String query = doPutPreparedStatementResultPOJO.getQuery();
+        final Connection connection = dataSource.getConnection();
+        final PreparedStatement statement = connection.prepareStatement(query,
+                ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
+
+        try (ArrowFileReader reader = new ArrowFileReader(new SeekableReadChannel(
+                new ByteArrayReadableSeekableByteChannel(
+                        doPutPreparedStatementResultPOJO.getParameters())), rootAllocator)) {
+
+          for (ArrowBlock arrowBlock : reader.getRecordBlocks()) {
+            reader.loadRecordBatch(arrowBlock);
+            VectorSchemaRoot vectorSchemaRootRecover = reader.getVectorSchemaRoot();
+            JdbcParameterBinder binder = JdbcParameterBinder.builder(statement, vectorSchemaRootRecover)
+                    .bindAll().build();
+
+            while (binder.next()) {
+              executeQuery(statement, listener);
+            }
+          }
+        }
+      } catch (StreamCorruptedException e) {
+        // Case where there are no parameters
+        final String query = new String(command.getPreparedStatementHandle().toString("UTF-8"));
+        final Connection connection = dataSource.getConnection();
+        final PreparedStatement preparedStatement = connection.prepareStatement(query,
+                ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);
+        executeQuery(preparedStatement, listener);
+      }
+    } catch (final SQLException | IOException | ClassNotFoundException e) {
+      LOGGER.error(format("Failed to getStreamPreparedStatement: <%s>.", e.getMessage()), e);
+      listener.error(CallStatus.INTERNAL.withDescription("Failed to prepare statement: " + e).toRuntimeException());
+    } finally {
+      listener.completed();
+    }
+  }
+
+  private void executeQuery(PreparedStatement statement,
+                            final ServerStreamListener listener) throws IOException, SQLException {
+    try (final ResultSet resultSet = statement.executeQuery()) {
+      final Schema schema = jdbcToArrowSchema(resultSet.getMetaData(), DEFAULT_CALENDAR);
+      try (final VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(schema, rootAllocator)) {
+        final VectorLoader loader = new VectorLoader(vectorSchemaRoot);
+        listener.start(vectorSchemaRoot);
+
+        final ArrowVectorIterator iterator = sqlToArrowVectorIterator(resultSet, rootAllocator);
+        while (iterator.hasNext()) {
+          final VectorSchemaRoot batch = iterator.next();
+          if (batch.getRowCount() == 0) {
+            break;
+          }
+          final VectorUnloader unloader = new VectorUnloader(batch);
+          loader.load(unloader.getRecordBatch());
+          listener.putNext();
+          vectorSchemaRoot.clear();
```

Comment:
Is this needed?

Reply:
Yes, it is needed. The root grows with each iteration if it is not cleared.